File: packages/altimate-engine/src/altimate_engine/connectors/mysql.py
"""MySQL warehouse connector with password and SSL authentication."""

from __future__ import annotations

from typing import Any

from altimate_engine.connectors.base import Connector


class MySQLConnector(Connector):
    """MySQL connector using mysql-connector-python SDK.

    Supports:
    - Password authentication
    - SSL connections
    """

    def __init__(
        self,
        host: str = "localhost",
        port: int = 3306,
        database: str | None = None,
        user: str | None = None,
        password: str | None = None,
        ssl_ca: str | None = None,
        ssl_cert: str | None = None,
        ssl_key: str | None = None,
        **kwargs,
    ):
        self.host = host
        self.port = port
        self.database = database
        self.user = user
        self.password = password
        self.ssl_ca = ssl_ca
        self.ssl_cert = ssl_cert
        self.ssl_key = ssl_key
        self.kwargs = kwargs
        self._conn = None

    def connect(self) -> Any:
        try:
            import mysql.connector
        except ImportError:
            raise ImportError(
                "mysql-connector-python not installed. Install with: pip install mysql-connector-python"
            )

        connect_params = {
            "host": self.host,
            "port": self.port,
            "user": self.user,
            "password": self.password,
        }
        if self.database:
            connect_params["database"] = self.database

        # SSL configuration
        if self.ssl_ca:
            connect_params["ssl_ca"] = self.ssl_ca
        if self.ssl_cert:
            connect_params["client_cert"] = self.ssl_cert
        if self.ssl_key:
            connect_params["client_key"] = self.ssl_key

        connect_params.update(self.kwargs)
        self._conn = mysql.connector.connect(**connect_params)
        return self._conn

    def _ensure_conn(self):
        if self._conn is None:
            self.connect()
        return self._conn

    def execute(
        self, sql: str, params: tuple | list | None = None, limit: int = 1000
    ) -> list[dict[str, Any]]:
        conn = self._ensure_conn()
        cur = conn.cursor(dictionary=True)
        cur.execute(sql, params)

        if cur.description is None:
            conn.commit()
            cur.close()
            return []

        rows = cur.fetchmany(limit)
        result = [dict(row) for row in rows]
        cur.close()
        return result

    def list_schemas(self) -> list[str]:
        rows = self.execute("SELECT schema_name FROM information_schema.schemata")
        return [row["schema_name"] for row in rows]

    def list_tables(self, schema: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = %s",
            (schema,),
        )
        return [{"name": row["table_name"], "type": row["table_type"]} for row in rows]

    def describe_table(self, schema: str, table: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT column_name, column_type AS data_type, is_nullable "
            "FROM information_schema.columns "
            "WHERE table_schema = %s AND table_name = %s ORDER BY ordinal_position",
            (schema, table),
        )
        return [
            {
                "name": row["column_name"],
                "data_type": row["data_type"],
                "nullable": row["is_nullable"] == "YES",
            }
            for row in rows
        ]

    def set_statement_timeout(self, timeout_ms: int) -> None:
        timeout_sec = max(1, timeout_ms // 1000)
        conn = self._ensure_conn()
        cur = conn.cursor()
        cur.execute(f"SET max_execution_time = {timeout_sec * 1000}")
        cur.close()

    def close(self) -> None:
        if self._conn is not None:
            self._conn.close()
            self._conn = None

File: packages/drivers/src/mysql.ts
/**
 * MySQL driver using the `mysql2` package.
 */

import type { ConnectionConfig, Connector, ConnectorResult, ExecuteOptions, SchemaColumn } from "./types"

export async function connect(config: ConnectionConfig): Promise<Connector> {
  let mysql: any
  try {
    mysql = await import("mysql2/promise")
    mysql = mysql.default || mysql
  } catch {
    throw new Error("MySQL driver not installed. Run: npm install mysql2")
  }

  let pool: any

  return {
    async connect() {
      const poolConfig: Record<string, unknown> = {
        host: config.host ?? "127.0.0.1",
        port: config.port ?? 3306,
        database: config.database,
        user: config.user,
        password: config.password,
        waitForConnections: true,
        connectionLimit: 5,
        connectTimeout: 10000,
      }

      if (config.ssl !== undefined) {
        poolConfig.ssl = config.ssl
      } else if (config.ssl_ca || config.ssl_cert || config.ssl_key) {
        const sslObj: Record<string, unknown> = {}
        if (config.ssl_ca) sslObj.ca = config.ssl_ca
        if (config.ssl_cert) sslObj.cert = config.ssl_cert
        if (config.ssl_key) sslObj.key = config.ssl_key
        poolConfig.ssl = sslObj
      }

      pool = mysql.createPool(poolConfig)
    },

    async execute(sql: string, limit?: number, _binds?: any[], options?: ExecuteOptions): Promise<ConnectorResult> {
      const effectiveLimit = options?.noLimit ? 0 : (limit ?? 1000)
      let query = sql
      const isSelectLike = /^\s*(SELECT|WITH|VALUES)\b/i.test(sql)
      if (
        isSelectLike &&
        effectiveLimit &&
        !/\bLIMIT\b/i.test(sql)
      ) {
        query = `${sql.replace(/;\s*$/, "")} LIMIT ${effectiveLimit + 1}`
      }

      const [rows, fields] = await pool.query(query)
      const columns = fields?.map((f: any) => f.name) ?? []
      const rowsArr = Array.isArray(rows) ? rows : []
      const truncated = effectiveLimit > 0 && rowsArr.length > effectiveLimit
      const limitedRows = truncated
        ? rowsArr.slice(0, effectiveLimit)
        : rowsArr

      return {
        columns,
        rows: limitedRows.map((row: any) =>
          columns.map((col: string) => row[col]),
        ),
        row_count: limitedRows.length,
        truncated,
      }
    },

    async listSchemas(): Promise<string[]> {
      const [rows] = await pool.query("SHOW DATABASES")
      return (rows as any[]).map(
        (r) => r.Database ?? r.database ?? Object.values(r)[0],
      ) as string[]
    },

    async listTables(
      schema: string,
    ): Promise<Array<{ name: string; type: string }>> {
      const [rows] = await pool.query(
        `SELECT table_name, table_type
         FROM information_schema.tables
         WHERE table_schema = ?
         ORDER BY table_name`,
        [schema],
      )
      return (rows as any[]).map((r) => ({
        name: (r.TABLE_NAME ?? r.table_name) as string,
        type: (r.TABLE_TYPE ?? r.table_type) === "VIEW" ? "view" : "table",
      }))
    },

    async describeTable(
      schema: string,
      table: string,
    ): Promise<SchemaColumn[]> {
      const [rows] = await pool.query(
        `SELECT column_name, data_type, is_nullable
         FROM information_schema.columns
         WHERE table_schema = ?
           AND table_name = ?
         ORDER BY ordinal_position`,
        [schema, table],
      )
      return (rows as any[]).map((r) => ({
        name: (r.COLUMN_NAME ?? r.column_name) as string,
        data_type: (r.DATA_TYPE ?? r.data_type) as string,
        nullable: (r.IS_NULLABLE ?? r.is_nullable) === "YES",
      }))
    },

    async close() {
      if (pool) {
        await pool.end()
        pool = null
      }
    },
  }
}
